aten/src/ATen/native/mps/operations/Copy.mm

```diff
@@ -414,24 +437,41 @@ void copy_blit_mps(void* dst, const void* src, size_t size) {
   auto dst_byte_offset = dst.storage_offset() * dst.itemsize();
   id<MTLBuffer> destBuffer = __builtin_bit_cast(id<MTLBuffer>, dst.storage().data());
 
-  dispatch_sync(stream->queue(), ^() {
-    @autoreleasepool {
-      id<MTLCommandBuffer> commandBuffer = stream->commandBuffer();
-      id<MTLBlitCommandEncoder> blitEncoder = [commandBuffer blitCommandEncoder];
-
-      [blitEncoder copyFromBuffer:sourceBuffer
-                     sourceOffset:src_byte_offset
-                         toBuffer:destBuffer
-                destinationOffset:dst_byte_offset
-                             size:size];
-      [blitEncoder endEncoding];
-      if (non_blocking) {
-        stream->commit(true);
-      } else {
+  auto srcDType = getMPSDataType(src.scalar_type());
+  auto dstDType = getMPSDataType(dst.scalar_type());
+  auto srcShape = getMPSShape(src);
+  auto dstShape = getMPSShape(dst);
+  if (srcDType == dstDType) {
+    dispatch_sync(stream->queue(), ^() {
+      @autoreleasepool {
+        id<MTLCommandBuffer> commandBuffer = stream->commandBuffer();
+        id<MTLBlitCommandEncoder> blitEncoder = [commandBuffer blitCommandEncoder];
+        [blitEncoder copyFromBuffer:sourceBuffer
+                      sourceOffset:0
+                          toBuffer:destBuffer
+                 destinationOffset:0
+                              size:size];
+        [blitEncoder endEncoding];
         stream->commitAndWait();
       }
+    });
+  } else {
+    @autoreleasepool {
```

Comment:
Can we refactor the castOperation to a separate function. There is redundancy here.